stdlib/public/SwiftShims/RefCount.h

```diff
@@ -684,6 +707,24 @@ class RefCounts {
   void initForNotFreeing() {
     refCounts.store(RefCountBits(0, 2), std::memory_order_relaxed);
   }
+  
+  // Initialize for an object which will never deallocate.
+  void initImmortal() {
+    refCounts.store(RefCountBits(RefCountBits::Immortal), std::memory_order_relaxed);
+  }
+  
+  void setIsImmortal(bool immortal) {
+    auto oldbits = refCounts.load(SWIFT_MEMORY_ORDER_CONSUME);
+    if (oldbits.isImmortal()) {
+      return;
+    }
+    RefCountBits newbits;
+    do {
+      newbits = oldbits;
+      newbits.setIsImmortal(immortal);
+    } while (!refCounts.compare_exchange_weak(oldbits, newbits,
```

Comment:
Huh, why don’t you have to re-read oldbits within this loop? If lose the race with another thread, don’t you need to read the value the other thread set and then modify that? Otherwise it seems like the exchange will fail forever.

I see this same pattern later in the patch so I’m probably misunderstanding.

Reply:
This is definitely confusing! When `compare_exchange_weak` fails because the current value no longer matches the expected value, it reads the new current value into the first parameter. And because it's C++, in/out parameters look exactly like normal parameters. Docs link for the curious: https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange

Reply:
I spent several hours over the weekend freaking out about this :)

Reply:
Same thing the first time I saw it. Fortunately, it only took me a minute to remember why this is OK this time around.